### PR TITLE
Don't emit mixin forwarders as final

### LINKED ISSUE #11485: Final methods in traits are not compiled as final
In 2.13.0-RC1, if we write this simple code 

```scala
import java.lang.reflect.Modifier.isFinal

trait HaveFinalMethod {
  def normalMethod: String = "normal"
  final def finalMethod: String = "final"
}

class Child extends HaveFinalMethod

object Main {
  def main(args: Array[String]): Unit = {
    println(isFinal(classOf[HaveFinalMethod].getMethod("normalMethod").getModifiers)) //false
    println(isFinal(classOf[HaveFinalMethod].getMethod("finalMethod").getModifiers)) //false but should be true!

    println(isFinal(classOf[Child].getMethod("normalMethod").getModifiers)) //false
    println(isFinal(classOf[Child].getMethod("finalMethod").getModifiers))  //true (as it should be)
  }
}
```

Then we'll see how `finalMethod` in `HaveFinalMethod` is not actually marked as final, which, on top of being inconsistent with what happens when we test the same on `Child`, causes issues with frameworks that use reflection.

if we make `Child` a trait instead of a class, then the problem is also present when inspecting `finalMethod`.

```scala
trait HaveFinalMethod {
  def normalMethod: String = "normal"
  final def finalMethod: String = "final"
}

trait Child extends HaveFinalMethod

object Main {
  def main(args: Array[String]): Unit = {
    println(isFinal(classOf[HaveFinalMethod].getMethod("normalMethod").getModifiers)) //false
    println(isFinal(classOf[HaveFinalMethod].getMethod("finalMethod").getModifiers)) //false but should be true!

    println(isFinal(classOf[Child].getMethod("normalMethod").getModifiers)) //false
    println(isFinal(classOf[Child].getMethod("finalMethod").getModifiers))  //false but should be true!
  }
}
```

### PR DESCRIPTION
**Pull request by smarter**
[edit] reverted in https://github.com/scala/scala/pull/8037

They're emitted as bridge since #7843 and a final bridge is a weird
concept that never comes up in Java (in particular, according to scala/bug#11485
ByteBuddy assumes that it can always override bridges). So let's just
drop the "final" part which is not essential for interop.

Fixes scala/bug#11485.

Note: the final flag cannot be dropped before the backend because the
inliner relies on this flag being present to know what can be inlined.

### CODE REVIEW

**Comment by smarter** [comment]
> scala.SerializationStabilityTest.testAll
> java.io.InvalidClassException: scala.math.Ordering$Int$; local class incompatible: stream classdesc serialVersionUID = -8412871093094815037, local class serialVersionUID = 3030081066962029638

The test pass when I run it locally, does that mean that a re-STARR is needed ?

**Comment by lrytz** [comment]
> The test pass when I run it locally, does that mean that a re-STARR is needed ?

No re-starr is needed, just update the test. When you run locally, the test looks at the scala-library compiled by the starr compiler, so the `final` flags are still there. On CI we always do a bootstrap, so the test looks at the scala-library compiled by the first iteration's compiler which no longer emits the `final` flag.

The better solution would be to make sure we consistently have `@SerialVersionUID` annotations on serializable classes / objects, that would avoid deserializtion failures in changes like this that don't acutally break serialization compatibility.

**Comment by lrytz** [comment]
For the code change itself, it's not obvious :-) The JVM spec doesn't say anything about combining `ACC_FINAL` with `ACC_BRIDGE`. Looking at the bytecode generated by javac, I also see that bridges are not marked final.

So +1 from me. Assigning to Adriaan.

Side-note: I looked again at the blog post you mentioned in your previous PR (http://stas-blogspot.blogspot.com/2010/03/java-bridge-methods-explained.html). Changing the method `SampleFour.A.foo` to be `final`, causes the bridge not to be emitted, and then `SampleFour.C.class.getMethod("foo").invoke(new SampleFour.C());` crashes with an `IllegalAccessException` :-)

File: src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala [hunk 1]
```diff
@@ -695,6 +695,7 @@ abstract class BTypesFromSymbols[G <: Global](val global: G) extends BTypes {
     val finalFlag = (
            (sym.isFinal || isTopLevelModuleClass(sym))
         && !sym.enclClass.isTrait
+        && !sym.isBridge // Final bridges confuse some frameworks (scala/bug#11485)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala]: Maybe add a reminder that mixin forwarders are also marked `ACC_BRIDGE`

**Comment by smarter** [comment]
> No re-starr is needed, just update the test.

Done, but that means that the test fails locally which might confuse others.

**Comment by smarter** [comment]
Hmm, it looks like the artifacts published on https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots are not compiled with a bootstrapped compiler since they still contain final bridges in the standard library which cause bytebuddy to blow up: https://github.com/mockito/mockito-scala/issues/95, so a re-STARR looks mandatory to check if everything is working correctly. But I'm surprised that the published snapshots are not bootstrapped, is there a reason for that ?

**Review by adriaanm — APPROVED** [review]

**Comment by lrytz** [comment]
Reverted in https://github.com/scala/scala/pull/8037

**Comment by dwijnand** [comment]
Reopen https://github.com/scala/bug/issues/11485 ?

**Comment by smarter** [comment]
No need, the underlying problem doesn't appear with the other revert.

**Pull request merged**